pkg/lightning/restore/check_info.go

```diff
@@ -0,0 +1,512 @@
+			msgs = append(msgs, fmt.Sprintf("compare columns perm failed. table %s's info has changed,"+
+				"consider remove this checkpoint, and start import again.", uniqueName))
+		}
+	}
+	return msgs, false, nil
+}
+
+// hasDefault represents col has default value.
+func hasDefault(col *model.ColumnInfo) bool {
+	return col.DefaultIsExpr || col.DefaultValue != nil || !mysql.HasNotNullFlag(col.Flag) ||
+		col.IsGenerated() || mysql.HasAutoIncrementFlag(col.Flag)
+}
+
+// SchemaIsValid checks the import file and cluster schema is match.
+func (rc *Controller) SchemaIsValid(ctx context.Context, tableInfo *md.MDTableMeta) ([]string, error) {
+	msgs := make([]string, 0)
+	info := rc.dbInfos[tableInfo.DB].Tables[tableInfo.Name]
+	if info != nil {
+		tableIgnoreColsMap := rc.cfg.Mydumper.IgnoreColumns
+		uniqueName := common.UniqueTable(tableInfo.DB, tableInfo.Name)
+		igCols := make(map[string]struct{})
+		if cols, ok := tableIgnoreColsMap[uniqueName]; ok {
+			for _, col := range cols {
+				igCols[col] = struct{}{}
+			}
+		}
+
+		if len(tableInfo.DataFiles) == 0 {
+			log.Info("no data files detected", zap.String("db", tableInfo.DB), zap.String("table", tableInfo.Name))
+			return nil, nil
+		}
+		// get columns name from data file.
+		dataFileMeta := tableInfo.DataFiles[0].FileMeta
+		var reader storage.ReadSeekCloser
+		var err error
+		if dataFileMeta.Type == md.SourceTypeParquet {
+			reader, err = md.OpenParquetReader(ctx, rc.store, dataFileMeta.Path, dataFileMeta.FileSize)
+		} else {
+			reader, err = rc.store.Open(ctx, dataFileMeta.Path)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		var parser md.Parser
+		blockBufSize := int64(rc.cfg.Mydumper.ReadBlockSize)
+		switch dataFileMeta.Type {
+		case md.SourceTypeCSV:
+			hasHeader := rc.cfg.Mydumper.CSV.Header
+			parser = md.NewCSVParser(&rc.cfg.Mydumper.CSV, reader, blockBufSize, rc.ioWorkers, hasHeader)
+		case md.SourceTypeSQL:
+			parser = md.NewChunkParser(rc.cfg.TiDB.SQLMode, reader, blockBufSize, rc.ioWorkers)
+		case md.SourceTypeParquet:
+			parser, err = md.NewParquetParser(ctx, rc.store, reader, dataFileMeta.Path)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		default:
+			msgs = append(msgs, fmt.Sprintf("file '%s' with unknown source type '%s'", dataFileMeta.Path, dataFileMeta.Type.String()))
+		}
+		err = parser.ReadRow()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		colsFromDataFile := parser.Columns()
+		colCountFromDataFile := len(parser.LastRow().Row)
+
+		colCountFromTiDB := len(info.Core.Columns)
+		colsFromTiDB := make([]string, 0, colCountFromTiDB)
+		defaultCols := make(map[string]struct{})
+		for _, col := range info.Core.Columns {
+			if hasDefault(col) || (info.Core.ContainsAutoRandomBits() && mysql.HasPriKeyFlag(col.Flag)) {
+				// this column has default value or it's auto random id, so we can ignore it
+				defaultCols[col.Name.L] = struct{}{}
+			}
+			colsFromTiDB = append(colsFromTiDB, col.Name.L)
+		}
```

Comment:
We may also do following checks for the ignored columns:
- if the column name not in target table schema report a warn
- if the column in target table has no default value, report an error